Add an initial binary protocol doc

bridadan opened the pull request:
This (at least partially) implements #22. This is a fairly rough document that goes over the binary protocol that I've been able to reverse engineer when writing [libsnapcast](https://github.com/bridadan/libsnapcast).

At this point I'm just hoping for an initial review (especially by @badaix) of the different fields since I don't know what some of them do (yet :wink:). This document has a bit of a bias since I'm only focusing on writing a client, not a server. Happy to extend this more or allow other people take over some of the server-focused bits!

bridadan commented:
Oops, thanks for pointing it at the right branch, I've rebased it on top of `develop`.

badaix commented on doc/binary_protocol.md:
```diff
@@ -0,0 +1,129 @@
+
+| Field          | Type    | Description                                                            |
+|----------------|---------|------------------------------------------------------------------------|
+| timestamp.sec  | int32   | The second value of the latency between the server and the client      |
```
this field is called `latency`

badaix commented:
thanks, it's already a quite good description of the protocol. I've added some comments, can you please check the review?

bridadan commented on doc/binary_protocol.md:
```diff
@@ -0,0 +1,129 @@
+
+1. Client opens a TCP socket to the server (default port is 1704)
+1. Client sends a [Hello](#hello) message
+    1. After sending this message, the client should be prepared to receive any typed message, not just the ones described below (time messages seem to come pretty consistently)
```
Thanks! I was getting confused looking at a WireShark capture I had made of the network traffic, I'll just drop this line.

bridadan commented on doc/binary_protocol.md:
```diff
@@ -0,0 +1,129 @@
+1. Server sends a [Stream Tags](#stream-tags) message
+1. Server sends a [Codec Header](#codec-header) message
+    1. Until the server sends this, the client shouldn't play any [Wire Chunk](#wire-chunk) messages
+1. The server will now send [Wire Chunk](#wire-chunk) and [Time](#time) messages consistently.
```
Awesome! This actually a really helpful explanation for when I start the time synchronization work. Once I have some kind of implementation working I'll feel more comfortable writing this piece, so keep an eye out for this in the future!

bridadan commented on doc/binary_protocol.md:
```diff
@@ -0,0 +1,129 @@
+| type                  | uint16 | Should be one of the typed message IDs                                                          |
+| id                    | uint16 | Not sure                                                                                        |
+| refersTo              | uint16 | I think this is used for responding to requests?                                                |
+| received.sec          | int32  | The second value of the timestamp when this message was received. Filled in by the client.      |
```
Good point! You can see my bias towards the client side of things :laughing:

bridadan commented on doc/binary_protocol.md:
```diff
@@ -0,0 +1,129 @@
+
+| Typed Message ID | Name                                 | Notes                                                                     |
+|------------------|--------------------------------------|---------------------------------------------------------------------------|
+| 0                | [Base](#base)                        | Shouldn't ever receive this as a value in base message?                   |
```
Great, I'll just drop that comment then.

bridadan commented on doc/binary_protocol.md:
```diff
@@ -0,0 +1,129 @@
+
+| Field          | Type    | Description                                                                           |
+|----------------|---------|---------------------------------------------------------------------------------------|
+| timestamp.sec  | int32   | The second value of the timestamp when this part of the stream should be played.      |
```
Ah nice, that helps me later on when I get synchronized playback working :smile:

bridadan commented:
Thanks for the quick response! I've addressed your comments, let me know if you see anything else. If you'd prefer to hold off on merging this until its more complete then no worries. But hopefully people find it useful.

badaix commented:
Thanks! Of course this is mergeable, no one can expect a project with version number < 1.0 to be complete in all regards :sweat_smile:
